File: cpp/0559.0_Maximum_Depth_of_N-ary_Tree.cpp
/*
DFS

执行用时：12 ms, 在所有 C++ 提交中击败了94.33% 的用户
内存消耗：10.4 MB, 在所有 C++ 提交中击败了93.79% 的用户
通过测试用例：38 / 38
*/

/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    int maxDepth(Node* root) {
        if (root == nullptr) {
            return 0;
        }
        int max_child_depth = 0;
        for (auto child: root->children) {
            max_child_depth = max(max_child_depth, maxDepth(child));
        }
        return max_child_depth + 1;
    }
};

File: leetcode-cn/0559.0_Maximum_Depth_of_N-ary_Tree.py
'''
DFS

执行用时：52 ms, 在所有 Python3 提交中击败了28.79% 的用户
内存消耗：16.6 MB, 在所有 Python3 提交中击败了83.43% 的用户
通过测试用例：38 / 38
'''
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def maxDepth(self, root: 'Node') -> int:
        ans = 0

        def dfs(node: 'Node', depth):
            nonlocal ans
            if not node:
                return
            ans = max(ans, depth + 1)
            for child in node.children:
                dfs(child, depth + 1)

        dfs(root, 0)
        return ans

'''
BFS

执行用时：40 ms, 在所有 Python3 提交中击败了89.50% 的用户
内存消耗：16.6 MB, 在所有 Python3 提交中击败了83.43% 的用户
通过测试用例：38 / 38
'''
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def maxDepth(self, root: 'Node') -> int:
        if not root:
            return 0

        step = 0
        q = deque([root])
        while q:
            for _ in range(len(q)):
                node = q.popleft()
                for child in node.children:
                    q.append(child)
            step += 1
        return step 

